aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -85,8 +87,20 @@ Tensor& _clamp_min_out_cpu(Tensor& result, const Tensor& self, Scalar min) {
   return legacy::th::_th_clamp_min_out(result, self, min);
 }
 
-Tensor& fill_(Tensor& self, Scalar value) {
-  return at::legacy::th::_th_fill_(self, value);
+Tensor& _fill__cpu(Tensor& self, Scalar value) {
+  auto iter = TensorIterator::unary_op(self, self);
+  if (iter->dtype() == at::ScalarType::Half) {
+    at::Half fill_value = value.to<at::Half>();
+    unary_kernel(*iter, [=](at::Half a) -> at::Half { return fill_value; });
+    return self;
+  }
+  if (iter->dtype() == at::ScalarType::Bool) {
+    // TODO: Replace this with TensorIterator
+    // It is undefined behavior to load values other than 0 or 1.
```

Comment:
This can't be undefined behavior. It should coerce `value` to 0/1.

Reply:
It causes undefined behavior within Unary loop to pass in a BoolTensor as self. See the [test failure](https://circleci.com/gh/pytorch/pytorch/1333021?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link).

Reply:
That's because you were doing `value.to<bool>()`. That's not what I mean by coerce. I mean something like:

```
bool fill_value;
if (value.isFloatingPoint())
  fill_value = (bool)value.to<double>();
else if (value.isIntegral())
  fill_value = (bool)value.to<int64_t>();
else
  <handle complex???>
```

i.e. the behavior should be well-defined if the user passes in something other than 0 or 1.

Reply:
Or maybe not -- I might be misunderstanding the issue here

Reply:
Yeah I was completely wrong here. The issue is the "unary_kernel" with zero Tensor arguments.